weave/actions_worker/tasks.py

```diff
@@ -0,0 +1,164 @@
+    ActionScore,
+)
+from weave.trace_server.refs_internal import InternalCallRef
+from weave.trace_server.trace_server_interface import (
+    ActionsAckBatchReq,
+    CallSchema,
+    CallsFilter,
+    CallsQueryReq,
+    FeedbackCreateReq,
+    RefsReadBatchReq,
+)
+
+WEAVE_ACTION_EXECUTOR_PACEHOLDER_ID = "WEAVE_ACTION_EXECUTOR"
+
+
+def ack_on_clickhouse(ctx: TaskCtx, succeeded: bool) -> None:
+    project_id = ctx["project_id"]
+    call_id = ctx["call_id"]
+    id = ctx["id"]
+    ClickHouseTraceServer.from_env().actions_ack_batch(
+        ActionsAckBatchReq(
+            project_id=project_id, call_ids=[call_id], id=id, succeeded=succeeded
+        )
+    )
+
+
+def publish_results_as_feedback(
+    ctx: TaskCtx, result: Any, configured_action_ref: str
+) -> None:
+    project_id = ctx["project_id"]
+    call_id = ctx["call_id"]
+    id = ctx["id"]
+    call_ref = InternalCallRef(project_id, call_id).uri()
+    ClickHouseTraceServer.from_env().feedback_create(
+        FeedbackCreateReq(
+            project_id=project_id,
+            weave_ref=call_ref,
+            feedback_type=base_model_name(ActionScore),
+            payload=ActionScore(
+                configured_action_ref=configured_action_ref, output=result
+            ).model_dump(),
+            wb_user_id=WEAVE_ACTION_EXECUTOR_PACEHOLDER_ID,
+        )
+    )
+
+
+def resolve_action_ref(configured_action_ref: str) -> ConfiguredAction:
+    server = ClickHouseTraceServer.from_env()
+    action_dict_res = server.refs_read_batch(
+        RefsReadBatchReq(refs=[configured_action_ref])
+    )
+    action_str = action_dict_res.vals[0]
+    action = ConfiguredAction.model_validate(json.loads(action_str))
+    return action
+
+
+def resolve_call(ctx: TaskCtx) -> CallSchema:
+    project_id, call_id = ctx["project_id"], ctx["call_id"]
+    server = ClickHouseTraceServer.from_env()
+    calls_query_res = server.calls_query(
+        CallsQueryReq(
+            project_id=project_id, filter=CallsFilter(call_ids=[call_id]), limit=1
+        )
+    )
+    return calls_query_res.calls[0]
+
+
+ActionConfigT = TypeVar("ActionConfigT")
+ActionResultT = TypeVar("ActionResultT")
+
+
+def action_task(
+    func: Callable[[str, str, ActionConfigT], ActionResultT],
+) -> Callable[[TaskCtx, str, str, str, ActionConfigT], ActionResultT]:
+    @wraps(func)
+    def wrapper(
+        ctx: TaskCtx,
+        call_input: str,
+        call_output: str,
+        configured_action_ref: str,
+        configured_action: ActionConfigT,
+    ) -> ActionResultT:
+        success = True
+        try:
+            scorer_name = func.__name__
+            result = func(call_input, call_output, configured_action)
+            publish_results_as_feedback(ctx, result, configured_action_ref)
+            print(f"Successfully ran {func.__name__}")
+            print(f"Result: {result}")
+        except Exception as e:
+            success = False
+            raise e
```

Comment:
What happens when error occurs? What is the impact of this raise on the celery worker?

Reply:
need to test but i think it results in the queue item not being ack'd, which would mean it could get reprocessed later.